aioconsole/stream.py

```diff
@@ -125,20 +126,55 @@ def __init__(self, stream, *, loop=None):
             loop = asyncio.get_event_loop()
         self.loop = loop
         self.stream = stream
+        self.buffer = deque()
+        self.write_task = None
 
     def write(self, data):
+        if self.stream is None:
+            raise RuntimeError("This writer stream is already closed")
         if isinstance(data, bytes):
             data = data.decode()
-        self.stream.write(data)
+        self.buffer.append(data)
+        if self.write_task is not None and not self.write_task.done():
+            return
+        if self.write_task is not None:
+            write_task, self.write_task = self.write_task, None
+            write_task.result()
+        self.write_task = asyncio.create_task(self._write_task_target(self.stream))
+
+    async def _write_task_target(self, stream):
+        while self.buffer:
+            data = self.buffer.popleft()
+            await self.loop.run_in_executor(None, stream.write, data)
 
     async def drain(self):
+        if self.write_task is not None:
+            try:
+                await self.write_task
+            finally:
+                self.write_task = None
         try:
             flush = self.stream.flush
         except AttributeError:
             pass
         else:
             await self.loop.run_in_executor(None, flush)
 
+    def close(self):
+        self.stream = None
+
+    def is_closing(self):
+        return self.stream is None and self.write_task is not None
+
+    async def wait_closed(self):
+        await self.drain()
+
+    def __del__(self):
+        if self.write_task is not None and not self.write_task.done():
+            self.write_task.cancel()
+        elif self.write_task is not None:
+            self.write_task.result()
```

Comment:
Seems like  `weakref.finalize` could be more robust approach than overriding `__del__()`, as long as `__del__()` had some problems with reference cycles, and, perhaps, it exposes some implementation-dependent behavior.

> However, handling of __del__() methods is notoriously implementation specific, since it depends on internal details of the interpreter’s garbage collector implementation.

So probably it's better to use `weakref.finalize()`, and just make finalizer right after creating the task?
```python
import asyncio
import weakref

async def count_forever():
    try:
        n = 0
        while True:
            await asyncio.to_thread(print, f'Still running: {n}')
            await asyncio.sleep(1.0)
            n += 1
    finally:
        await asyncio.to_thread(print, f'Task stopped successfully: {n}')


def finalize_task(task):
    if not task.done():
        task.cancel()
    else:
        task.result()

class Stream:

    def write(self, data):
        ...
        task = asyncio.create_task(count_forever())
        weakref.finalize(self, finalize_task, task)


for _ in range(2):
    s = Stream()
    s.write(...)
    await asyncio.sleep(2)
    del s
```
---
Output:
```
Still running: 0
Still running: 1
Task stopped successfully: 1
Still running: 0
Still running: 1

Task stopped successfully: 1
```
@vxgmichel , what do you think?

Reply:
I didn't know about `weakref.finalize` but it looks interesting. Maybe it will help with the pypy tests in the CI.

Here's my attempt: https://github.com/vxgmichel/aioconsole/pull/79/commits/06a4abc9f1fbf845b485cf695fbd7fd38ec346ec

Reply:
Seems like my last commit actually fixed the CI: bc4aa0a

Feel free to review the PR if you want to :)

Reply:
lgtm :)